sycl/source/detail/graph_impl.hpp

```diff
@@ -0,0 +1,362 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/detail/cg_types.hpp>
+#include <sycl/detail/os_util.hpp>
+#include <sycl/ext/oneapi/experimental/graph.hpp>
+#include <sycl/handler.hpp>
+
+#include <detail/kernel_impl.hpp>
+
+#include <cstring>
+#include <functional>
+#include <list>
+#include <set>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+
+namespace ext {
+namespace oneapi {
+namespace experimental {
+namespace detail {
+
+/// Implementation of node class from SYCL_EXT_ONEAPI_GRAPH.
+class node_impl {
+public:
+  /// List of successors to this node.
+  std::vector<std::shared_ptr<node_impl>> MSuccessors;
+  /// List of predecessors to this node.
+  ///
+  /// Using weak_ptr here to prevent circular references between nodes.
+  std::vector<std::weak_ptr<node_impl>> MPredecessors;
+  /// Type of the command-group for the node.
+  sycl::detail::CG::CGTYPE MCGType = sycl::detail::CG::None;
+  /// Command group object which stores all args etc needed to enqueue the node
+  std::unique_ptr<sycl::detail::CG> MCommandGroup;
+
+  /// True if an empty node, false otherwise.
+  bool MIsEmpty = false;
+
+  /// Add successor to the node.
+  /// @param Node Node to add as a successor.
+  /// @param Prev Predecessor to \p node being added as successor.
+  ///
+  /// /p Prev should be a shared_ptr to an instance of this object, but can't
+  /// use a raw \p this pointer, so the extra \Prev parameter is passed.
+  void register_successor(const std::shared_ptr<node_impl> &Node,
+                          const std::shared_ptr<node_impl> &Prev) {
+    MSuccessors.push_back(Node);
+    Node->register_predecessor(Prev);
+  }
+
+  /// Add predecessor to the node.
+  /// @param Node Node to add as a predecessor.
+  void register_predecessor(const std::shared_ptr<node_impl> &Node) {
+    MPredecessors.push_back(Node);
+  }
+
+  /// Construct an empty node.
+  node_impl() : MIsEmpty(true) {}
+
+  /// Construct a node representing a command-group.
+  /// @param CGType Type of the command-group.
+  /// @param CommandGroup The CG which stores the command information for this
+  /// node.
+  node_impl(sycl::detail::CG::CGTYPE CGType,
+            std::unique_ptr<sycl::detail::CG> &&CommandGroup)
+      : MCGType(CGType), MCommandGroup(std::move(CommandGroup)) {}
+
+  /// Recursively add nodes to execution stack.
+  /// @param NodeImpl Node to schedule.
+  /// @param Schedule Execution ordering to add node to.
+  void topology_sort(std::shared_ptr<node_impl> NodeImpl,
+                     std::list<std::shared_ptr<node_impl>> &Schedule) {
+    for (auto Next : MSuccessors) {
+      // Check if we've already scheduled this node
+      if (std::find(Schedule.begin(), Schedule.end(), Next) == Schedule.end())
```

Comment:
`n` is expected to be fairly small yes. Also might be worth noting this is done in the call to `finalize()` where I think we would consider optimization less critical. Since in the ideal case you would finalize the graph once and then never have to do it again.